snappy/binaries.go

```diff
@@ -30,22 +30,33 @@ import (
 	"github.com/ubuntu-core/snappy/arch"
 	"github.com/ubuntu-core/snappy/dirs"
 	"github.com/ubuntu-core/snappy/osutil"
+	"github.com/ubuntu-core/snappy/snap"
 	"github.com/ubuntu-core/snappy/snap/snapenv"
 )
 
-// generate the name
-func generateBinaryName(m *snapYaml, app *AppYaml) string {
+func oldGenerateBinaryName(m *snapYaml, app *AppYaml) string {
 	var binName string
 	if app.Name == m.Name {
 		binName = filepath.Base(app.Name)
 	} else {
 		binName = fmt.Sprintf("%s.%s", m.Name, filepath.Base(app.Name))
 	}
+	return filepath.Join(dirs.SnapBinariesDir, binName)
+}
+
+// generate the name
+func generateBinaryName(app *snap.AppInfo) string {
+	var binName string
+	if app.Name == app.Snap.Name {
+		binName = filepath.Base(app.Name)
+	} else {
+		binName = fmt.Sprintf("%s.%s", app.Snap.Name, filepath.Base(app.Name))
+	}
 
 	return filepath.Join(dirs.SnapBinariesDir, binName)
 }
 
-func binPathForBinary(pkgPath string, app *AppYaml) string {
+func binPathForBinary(pkgPath string, app *snap.AppInfo) string {
```

Comment:
Might become AppInfo.CommandLine.

Note that the name is misleading. app.Command holds a command with arguments, not a path.